sched/scheduler/stateful_scheduler.go

```diff
@@ -351,25 +353,41 @@ func (s *statefulScheduler) scheduleTasks() {
 		}
 
 		s.asyncRunner.RunAsync(
-			runner.run,
+			run.run,
 			func(err error) {
 				flaky := false
 				if err != nil {
 					// Get the type of error. Currently we only care to distinguish runner (ex: thrift) errors to mark flaky nodes.
 					taskErr := err.(*taskError)
 					flaky = (taskErr.runnerErr != nil)
-					if !preventRetries {
-						log.Info("Error running job:", jobId, ", task:", taskId, " cmd:", taskDef.Argv, "(will be retried)", " err:", err)
-						jobState.errorRunningTask(taskId, err)
-					} else {
-						log.Info("Error running job:", jobId, ", task:", taskId, " cmd:", taskDef.Argv, "(will not be retried)", " err:", err)
+
+					msg := "Error running job (will be retried):"
+					if taskErr.resultErr != nil && taskErr.st.State == runner.COMPLETE {
+						msg = "Error running job (quitting, tasks that run to completion are not retried):"
+						err = nil
+					} else if preventRetries {
+						msg = fmt.Sprintf("Error running job (quitting, hit max retries of %d):", s.maxRetriesPerTask)
 						err = nil
+					} else {
+						jobState.errorRunningTask(taskId, err)
+					}
+					log.Info(msg, jobId, ", task:", taskId, " err:", taskErr, " cmd:", taskDef.Argv)
+
+					// If the task completed succesfully but sagalog failed, start a goroutine to retry until it succeeds.
+					if taskErr.sagaErr != nil && taskErr.runnerErr == nil && taskErr.resultErr == nil {
+						log.Info(msg, jobId, ", task:", taskId, " -> starting goroutine to handle failed sagalog. ")
+						go func() {
+							for err := errors.New(""); err != nil; err = run.logTaskStatus(&taskErr.st, saga.EndTask) {
```

Comment:
There are two options here as I see it - keep trying until we manually intervene, or panic since we can't communicate status to our clients. I'm doing the first thing here since it seems the better of the two (also it really doesn't matter for Q2 since we use in-memory sagalog).